core: Fix minor regression on colorless arches

jlebon opened the pull request:
In 11aeb825d2cb ("core: Handle RPM file colors more efficiently"), we
added an early exit in `handle_file_dispositions` on arches which don't
use colours. However, this is incorrect because that function does more
than just colour handling; it also handles files shared between RPMs.

Fix this and expand the function's docstring.

We have tests for this, but it passed through because CI runs on x86_64.
But this would matter or i686 or armhfp (which is relevant for Fedora
IoT).

cgwalters commented:
> But this would matter or i686 or armhfp (which is relevant for Fedora IoT).

Is there a related BZ or issue or conversation about this?

jlebon commented:
> > But this would matter or i686 or armhfp (which is relevant for Fedora IoT).
> 
> Is there a related BZ or issue or conversation about this?

No, was randomly thinking about https://github.com/coreos/rpm-ostree/pull/3161 yesterday while taking a shower and it just came to me.

jlebon commented:
Added a commit to adapt CI for f35! :arrow_up:

cgwalters commented on src/libpriv/rpmostree-core.cxx:
```diff
@@ -2503,10 +2503,15 @@ ht_insert_path_for_nevra (GHashTable *ht, const char *nevra, char *path, gpointe
+ *
+ * The librpm functions and APIs for these are unfortunately private since they're just run
+ * as part of rpmtsRun(). XXX: see if we can make the rpmfs APIs public, but we'd still need
+ * to support RHEL/CentOS anyway. */
```
It doesn't matter here but I'd s/RHEL/RHEL8/ at least mentally - remember "RHEL" is older until suddenly it isn't :smile:

jlebon replied:
It's an old comment, but yeah misleading now given CentOS Stream. Had to respin for CI so dropped that bit entirely!

cgwalters reviewed (approved):
> In 11aeb82 ("core: Handle RPM file colors more efficiently"), we
added an early exit in handle_file_dispositions on arches which don't
use colours. However, this is incorrect because that function does more
than just colour handling; it also handles files shared between RPMs.

Isn't this really "11aeb82 was wrong"?  This is *completely* fine as is, but one option here is to structure things like this in the future as "revert + new fix" to make that even clearer.

> No, was randomly thinking about #3161 yesterday while taking a shower and it just came to me.

Amazing :smile:

jlebon commented:
Bahh, copy/paste error in the f35 transition commit. Fixed that now!

> > In [11aeb82](https://github.com/coreos/rpm-ostree/commit/11aeb825d2cbac36429e035aa9efbaa522b995dc) ("core: Handle RPM file colors more efficiently"), we
> > added an early exit in handle_file_dispositions on arches which don't
> > use colours. However, this is incorrect because that function does more
> > than just colour handling; it also handles files shared between RPMs.
> 
> Isn't this really "11aeb82 was wrong"? This is _completely_ fine as is, but one option here is to structure things like this in the future as "revert + new fix" to make that even clearer.

To be clear, the crux of 11aeb82 was fine and doesn't need reverting. We still e.g. only add to the hash table those files that have colors, which is the major improvement. Not sure, might have been more confusing I think if we reverted everything and added it back without the problematic bit at the top?
